### PR TITLE
OCPBUGS-5943: Enable topology-aware hints if, and only if, nodes have zones

### PR DESCRIPTION
**Pull request by Miciah**
Only set the `service.kubernetes.io/topology-aware-hints` annotation on the DNS service if at least two nodes are ready and all nodes that are ready have allocatable CPU and have the `topology.kubernetes.io/zone` label.

This change reduces the risk that the endpointslice controller will emit warning events because the service specifies the annotation to enable the topology-aware hints feature but the cluster has insufficient information to enable the feature.  Without this change, a cluster on which nodes were lacking the `topology.kubernetes.io/zone` label would emit "TopologyAwareHintsDisabled" events repeatedly.

* `pkg/operator/controller/controller.go` (`New`): Watch nodes in case the controller needs to add or remove the `service.kubernetes.io/topology-aware-hints` annotation in response to a node event, using the new `ignoreNodeForTopologyAwareHints` and `nodeIsValidForTopologyAwareHints` helpers.
* `pkg/operator/controller/controller_dns_service.go` (`ensureDNSService`): Use the new `shouldEnableTopologyAwareHints` method and pass the result to `desiredDNSService`.
(`shouldEnableTopologyAwareHints`): New method.  Return a Boolean indicating whether the topology-aware hints feature should be enabled, using the new `ignoreNodeForTopologyAwareHints` and `nodeIsValidForTopologyAwareHints` helpers.
(`ignoreNodeForTopologyAwareHints`, `nodeIsValidForTopologyAwareHints`): New functions.
(`desiredDNSService`): Add an `enableTopologyAwareHints` parameter.  Only set the `service.kubernetes.io/topology-aware-hints` annotation on the service if `enableTopologyAwareHints` is true.
* `pkg/operator/controller/controller_dns_service_test.go` (`Test_shouldEnableTopologyAwareHints`): New test.  Verify that `shouldEnableTopologyAwareHints` behaves as expected.
(`fakeCache`): New type, used in `Test_shouldEnableTopologyAwareHints`.
* `go.mod`: 
* `go.sum`: 
* `vendor/*`: Regenerate.

### CODE REVIEW

**Comment by Miciah** [comment]
[e2e-aws-ovn-single-node passed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/364/pull-ci-openshift-cluster-dns-operator-master-e2e-aws-ovn-single-node/1650652904429719552), and as expected, the "dns-default" service does not have the `service.kubernetes.io/topology-aware-hints` annotation.

[e2e-aws-ovn-operator failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/364/pull-ci-openshift-cluster-dns-operator-master-e2e-aws-ovn-operator/1650652904379387904) because `TestDNSLogging` failed:
```
 === RUN   TestDNSLogging
    operator_test.go:325: failed to get NXDOMAIN entry for test.svc.cluster.local. which does not exist
--- FAIL: TestDNSLogging (806.17s) 
```
However, the "dns-default" service does have the expected annotation in [`services.json`](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/364/pull-ci-openshift-cluster-dns-operator-master-e2e-aws-ovn-operator/1650652904379387904/artifacts/e2e-aws-ovn-operator/gather-extra/artifacts/services.json), and [the dns-operator logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-dns-operator/364/pull-ci-openshift-cluster-dns-operator-master-e2e-aws-ovn-operator/1650652904379387904/artifacts/e2e-aws-ovn-operator/gather-extra/artifacts/pods/openshift-dns-operator_dns-operator-7875c4bc7b-cfx8h_dns-operator.log) show a single update to add the annotation.
/test e2e-aws-ovn-operator

File: pkg/operator/controller/controller_dns_service_test.go [hunk 1]
```diff
@@ -152,3 +162,113 @@ func TestDNSServiceChanged(t *testing.T) {
 		}
 	}
 }
+
+func Test_shouldEnableTopologyAwareHints(t *testing.T) {
```
Comment (frobware) [inline pkg/operator/controller/controller_dns_service_test.go:166]: Looking at the code coverage for `shouldEnableTopologyAwareHints` when this test is run I don't see test cases that cover:
- nodeHasExcludedLabels returns true
- !nodeIsReady
- node.Status.Allocatable.Cpu().IsZero() returns true

Comment (Miciah) [reply pkg/operator/controller/controller_dns_service_test.go:166]: https://github.com/openshift/cluster-dns-operator/compare/a343ca809da3079230527d06bf1815add9cb4afd..4a7037f7704ac2b0a44f9e22d3b28e42b24db5a6 adds test cases for those three conditions.

File: pkg/operator/controller/controller_dns_service_test.go [hunk 2]
```diff
@@ -152,3 +162,161 @@ func TestDNSServiceChanged(t *testing.T) {
...
+	emptyLabels := map[string]string{}
+	someCPU := map[corev1.ResourceName]resource.Quantity{
+		"cpu": resource.MustParse("1m"),
+	}
+	noCPU := map[corev1.ResourceName]resource.Quantity{
+		"cpu": resource.MustParse("0"),
```
[the remark this answers is not in the document]
Comment (frobware) [reply pkg/operator/controller/controller_dns_service_test.go:172]: Perhaps "expecting" is too strong a term; I fiddled with the resource name to see what the outcome would be.

File: pkg/operator/controller/controller_dns_service.go [hunk 3]
```diff
@@ -76,7 +82,65 @@ func (r *reconciler) currentDNSService(dns *operatorv1.DNS) (bool, *corev1.Servi
...
+// topology-aware hints can be enabled for the DNS service.
+//
+// Topology-aware hints should be enabled if, and only if, there are at least 2
+// ready nodes and they all have allocatable CPU and specify topology zones.
+//
+// Much of this logic is copied from
+// <https://github.com/openshift/kubernetes/blob/b40493584076fb1ab29f3bed1d05d16cbc5b17f1/pkg/controller/endpointslice/topologycache/topologycache.go#L203-L262>.
+func (r *reconciler) shouldEnableTopologyAwareHints(dns *operatorv1.DNS) (bool, error) {
+	var nodesList corev1.NodeList
+	if err := r.cache.List(context.TODO(), &nodesList); err != nil {
```
Comment (gcs278) [inline pkg/operator/controller/controller_dns_service.go:95]: So the dns control loop doesn't watch for node events, like creation, ready/not ready, label change, etc.

What happens when the reconcile loop runs, but then a node event happens? Say, adding a label for zone topology, or scale up event? Do we just have to wait for another reconcile event? Should we be watching nodes now?

File: pkg/operator/controller/controller.go [hunk 4]
```diff
@@ -90,15 +91,43 @@ func New(mgr manager.Manager, config operatorconfig.Config) (controller.Controll
...
+	nodePredicate := func(o client.Object) bool {
+		node := o.(*corev1.Node)
+		return !ignoreNodeForTopologyAwareHints(node)
+	}
+	if err := c.Watch(&source.Kind{Type: &corev1.Node{}}, handler.EnqueueRequestsFromMapFunc(objectToDNS), predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool { return nodePredicate(e.Object) },
+		DeleteFunc: func(e event.DeleteEvent) bool { return nodePredicate(e.Object) },
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			old := e.ObjectOld.(*corev1.Node)
+			new := e.ObjectOld.(*corev1.Node)
```
Comment (gcs278) [inline pkg/operator/controller/controller.go:119]: These are the same objects no?

Comment (Miciah) [reply pkg/operator/controller/controller.go]: Yeah.  Thanks!  Fixed in https://github.com/openshift/cluster-dns-operator/compare/781c2b876f81169e3461eaf7b724de68a7759ada..55550c70282fc27d55b99dddcb0b8ba49b73dcdc.

**Comment by gcs278** [comment]
/approve
@frobware want to look at the recent updates?

**Comment by frobware** [comment]
> /approve @frobware want to look at the recent updates?

/lgtm

**Pull request merged**